docs/design/2022-01-04-integer-shard-index.md

```diff
@@ -0,0 +1,591 @@
+| 17 14:50 | 800    | 3      | 3      | 4400    | 3000    | insert:  61     update:111   | 49     |        | 第二轮 |
+| 17 15:40 | 800    | 3      | 4      | 5000    | 3500    | insert:  57     update:103   | 53     | 19.30% | 第一轮 |
+| 17 9:47  | 800    | 3      | 4      | 5100    | 3600    | insert:  56     update:100   | 54     | 30.60% | 第二轮 |
+
+​																					hot index scalability
+
+
+
+| 测试时间 | 并发数 | tidb数 | tikv数 | tidbCPU | tikvCPU | duration                    | QPS  | 扩展性 | 备注   |
+| -------- | ------ | ------ | ------ | ------- | ------- | --------------------------- | ---- | ------ | ------ |
+| 17 18:50 | 800    | 3      | 3      | 5200    | 3100    | insert: 60     update:110   | 47.5 |        | 第一轮 |
+| 17 19:30 | 800    | 3      | 3      | 5200    | 3000    | insert:  60     update:110  | 47.2 |        | 第二轮 |
+| 17 20:10 | 800    | 3      | 4      | 6200    | 3800    | insert:  56.5     update:88 | 57.5 | 63.20% | 第一轮 |
+| 17 20:40 | 800    | 3      | 4      | 6250    | 3800    | insert:  56.5     update:88 | 58.5 | 71.40% | 第二轮 |
+
+​																					shard index scalability
+
+## Motivation
+
+We made a new built-in function "tidb_shard()" as the expression prefix of shard index. There is only one integer parameter for tidb_shard function. What index is a shard index? It must meet the follower rules.
+
+- It is an unqie index
+- The first field of index is tidb_shard(`a`) , the column `a` is integer type and is the second field of the index.
+
+When querying based on the original index field, the TiDB optimizer should automatically recognize the shard index and add the expression field `tidb_shard(a) = val` to the query condition.
+
+How to solve the hot index problem by shard index? For example, the definition of a table is `test(id int primary key clustered, a int, b int, unique key idx_a(a))` . The index `idx_a` is a hot index because of monotonically increasing data. We modify the definition of the index `idx_a` as follows `unique key idx_a((tidb_shard(a)),a)`. The written data of index `idx_a` is not monotonically increasing now. But here is a question, how the optimizer of TiDB chooses the index for original queries that are designed for `idx_a(a)`? e.g. `SELECT * FROM test WHERE a = 100`, it can't use the unique index 'idx_a((tidb_shard(a)),a' any more. We add a new rule for TiDB optimizer, if the access condition of a query contains all the fields of shard index except the first `tidb_shard(a)` expression, The optimizer adds the expression`tidb_shard(a) = xxx` to the access condition and the new query statement is like `SELECT * FROM test WHERE tidb_shard(a) = 8 AND a = 100`, so that the query can use the shard index.
+
+## Detailed design
+
+### tidb_shard function
+
+#### dsciription
+
+Only support an integer parameter. It transforms an integer number to a value in the range [0, 255]. e.g. `tidb_shard(100)` is the value `8` .
+
+#### principle
```

Comment:
We could demonstrate the processing process using some pseudo code or a flow graph, much code may make it difficult for the community developers to get the key ideas.

Reply:
I fixed (adder *exprPrefixAdder) addExprPrefix4DNFCond、AddExpr4EqAndInCondition、AddExpr4EqAndInCondition、NeedAddGcColumn4ShardIndex、AddGcColumn4EqCond、AddGcColumn4InCond，replaced the code to function process explanations. But the other functions, I think they are so simple that they can explain themselves by code